File: 28. 搜索二维矩阵.cpp
class Solution {
public:
    /**
     * @param matrix: matrix, a list of lists of integers
     * @param target: An integer
     * @return: a boolean, indicate whether matrix contains target
     */
    bool searchMatrix(vector<vector<int>> &matrix, int target) {
        if (matrix.empty()) return false;
        int row_end = matrix.size();
        int column_end = matrix[0].size();
        int row_index = 0;
        int column_index = column_end - 1;
        while (row_index < row_end && column_index >= 0) {
            if (target < matrix[row_index][column_index]) column_index -= 1;
            else if (target > matrix[row_index][column_index]) row_index += 1;
            else return true;
        }
        return false;
    }
};

// 总耗时 218 ms
// 您的提交打败了 78.80% 的提交!

File: Python3/28. 搜索二维矩阵.py
#!/usr/bin/env python3

class Solution:
    """
    @param matrix: matrix, a list of lists of integers
    @param target: An integer
    @return: a boolean, indicate whether matrix contains target
    """
    def searchMatrix(self, matrix, target):
        if matrix == [] or matrix == None : return False
        row = len(matrix)
        column = len(matrix[0])
        row_index = 0
        column_index = column - 1
        while row_index < row and column_index >= 0 :
            if target > matrix[row_index][column_index] : row_index += 1
            elif target < matrix[row_index][column_index] : column_index -= 1
            else : return True
        return False

# 总耗时 1017 ms
# 您的提交打败了 97.44% 的提交!